### DISCUSSION TITLE
5.6

### DISCUSSION THREAD

**Discussion by RomanKubiak** [RomanKubiak/ctrlr#145] [Releases]
I built and uploaded a new minor version 5.6 (it's a pre-release), it is a big change since it migrates to JUCE 6.0.7 from JUCE 5, so it's stability is a big question.
There are builds for Windows Mac (i did my build on mac Mojave 10.14.6) and Linux
let me know if it works or not for you

**Comment by dobo365** [comment]
Thx! I downloaded the Windows version and will make some tests this w-e with my existing panels.

**Comment by Ted-juh** [comment]
Cool. Going to give it a try as well. Does that mean that the added functions for certain existing classes are available with lua as well? 

On a sidenote, I know it probably isn't high priority, but is it possible to add certain juce classes? Like viewport or rectangleList for example?

**Reply by RomanKubiak** [reply]
i can try to add more classes, please open an issue on github with the classes you'd like added. i'll hvae a look how much is possible to add

**Comment by tomlem** [comment]
I will give the linux version a try.
Where can I find this build?

Edit: Ok, I had misunderstand the message. I've found the code.

**Comment by tomlem** [comment]
Trying to compile on debian buster a got this error:

----------------------------------------
```
~/build4/ctrlr/Scripts$ ./post-commit 
CTRLR pre-commit HOOK
CTRLR: build_now_linux
CTRLR: CTRLR_ROOT /home/tom/build4/ctrlr/Scripts/../

	Version: 5.6.2
	CtrlrRevision.template: /home/tom/build4/ctrlr/Scripts/..//Source/Core/CtrlrRevision.template
	CtrlrRevision.h: /home/tom/build4/ctrlr/Scripts/..//Source/Core/CtrlrRevision.h
	Date: zo jan 31 16:40:40 CET 2021

CTRLR: Delete binaries
CTRLR: check if boost is unpacked
CTRLR: boot directory exists
CTRLR: Build Standalone
CTRLR[linux]: Building for x86_64, JOBS 16
CTRLR[linux]: Compile PCH
stadfx.h
stdafx_luabind.h
CTRLR[linux]: Build now

Makefile:336: *** target pattern contains no '%'.  Stop.
CTRLR[linux]: build failed

```
----------------------------------------

**Reply by spiffo-smith** [reply]
Trying to Compile **Ctrlr_SharedCode** on **Visual Studio 2019** and got more or less the same Error:

\ctrlr\Source\UIComponents\CtrlrAbout.cpp(4,10): fatal error C1083: Cannot open include file: **'CtrlrRevision.h'**: No such file or directory

That File is an include in **CtrlrAbout.cpp** but doesn't seem to be anywhere in the Source Directory?

**Reply by damiensellier** [reply]
Hi, I followed @godlike-productions tutorial and never had this issue, I'm not sure but maybe the Cygwin64 step could help : 
https://godlike.com.au/fileadmin/godlike/techtools/ctrlr/guides/Compiling_Ctrlr_for_Windows_10_v2.1.pdf

**Reply by spiffo-smith** [reply]
Yes I’ve used that guide before when I compiled it successfully about a year ago, and I never had to run those scripts. What on earth are they for anyway and what could they have to do with what appears to be a simple case of a missing .h file?

**Reply by shooking** [reply]
Would be true story except Roman uses JUCE and that requires the source to be managed by Projucer. So it isn't a simple case of using git to modify files.
What I have never understood is why there were no historical release points - version of the branch that should build. It makes Ctrlr really tricky to contribute to. For example I wanted to add in Rapsberry Pi support.

Nowadays I use mido and TkInter on Python with a touchscreen on Pi4. Gives me what I need in a way I can control. No Projucer steps etc.

**Reply by damiensellier** [reply]
@shooking we are screwed, JUCE 7 is out :)

**Comment by dobo365** [comment]
OK. I have now installed 5.6 and made some tests but had to stop almost directly.

Comments:
- I prefer the old 5.3.201 look and feel
- properties values are better separated and the whole property list is more readable
- Lua editor - Methods tab is also better and I prefer the old layout than the new one with just one plain color
- Same for Ctrlr background. Prefer the old background than the new one with just one plain color
- I have the impression that the old layout is less tiring for the eyes. Difficult to express...

Tests and issues:
- I loaded at random one of my panels. It went OK. There are different things happening (initializing different types of modulators as variables, settings some uiLabels, loading a predefined patch stored in a memory block...)
- Load a patch from my PC disk: OK
- try to move a uiImageSlider with the mouse: NOK. Ctrlr crashes
- hover a uiImageSlider then use the scrollwheel to change value: NOK Ctrlr crashes

As this is rather basic, I propose that you try finding why this is crashing before I continue more tests with an existing panel.
I will make some tests with a new empty panel to see the situation.

**Comment by Ted-juh** [comment]
Big thumbs up for the search function. Really helpful when you have to change just one property for a lot of modulators. But, to enhance workflow even more:

- Is it possible to always show the Modulator name? When a modulator is locked, it's hard to keep track of which modulator was selected. 
- Add an "X" to the search bar to delete its contents and go back to full properties view right away. 

On the cosmetics side:
 I have to agree with Goodweather that, to avoid eyestrain, less white is better for the properties ID's/ Names. And the Blue background is eh... very blue. But I like the step to the more modern design from the "Windows 95" look.

Some Issues (or not):
- Didn't encounter the uiImageSlider problem.  
- The "X" to close a panel from the tabbed view is gone. The same goes for the Lua editor. Always have to go to the menu and choose "close panel" or "close current tab".

- When I close a panel it always says that there are unsaved changes with the options to "save", "discard" or "cancel" it. Even though I didn't do anything. In Ctrlr 5.3.201 it just asks whether I want to close the panel or not.

- My SH-201 panel doesn't load. It just closes Ctrlr without an error while it loads without problems in 5.3.201. When I load Ctrlr again it is reverted back to the panel I loaded before I loaded the SH-201 panel. But there is no log file. And yes, it is ticked in the Ctrlr preferences window. Now that could be due to my crappy coding but most panels with lots of methods attached to it seem to work fine.

- "LuaModulatorValueChange not updating or showing up at all in the properties in Panel Mode problem" is still present. I opened a topic about it on the forum.

**Comment by Possemo** [comment]
Do you guys test the panel with real hardware? Probably not. Or maybe you are using the Mac version? At least on Windows (and according to dehnhard also Linux) this build does not work at all as it won't let you select a MIDI-port. But never mind - a contributor called dehnhardt has already fixed it and has opened a pull request. Thus on the next release MIDI-ports will be enabled again.

I don't care a lot about the missing close button. I found it much more annoying that on the Lua Editor you don't see the cursor anymore. Apart from that I don't care much about "new design" or "old design". It's nice that you can select the "LookAndFeel scheme for the panel". This concerns maninly the uiButton. One bug on that: uiButtons change on all layers but not on layer 0 (background layer).

The missing close button could be a problem of the "Introjucer" legacy. I am experimenting in building Ctrlr, thus I also tried to do it from the start using Projucer. I then saw that when you generate procedures done with Introjucer it screws it all up. So you cannot reload them with Projucer or you get one big mess. I think sooner or later the code will have to be adapted to Projucer as Introjucer is gone since a long time (last Introjucer was bundled with JUCE v4.1 or similar). These Introjucer procedures will start to make problems...

**Reply by Ted-juh** [reply]
I couldn't test it with real hardware because the panel for my synth wouldn't load, simple as that. But Ctrlr showed all the midi settings being set to my synth without throwing an error, probably because it takes the settings from the same folder as the old version. 

Now I wondered why I didn't get an error since I am on windows 10 as well. But it seems that Windows has a habit of disabling unsigned drivers after a big update and I just updated last week. But after reinstalling the drivers for my synth and finding a version of my panel that wanted to load I have to confirm that the panel opens with an error: Warning [Error  ][Some timestamp]CtrlrMIDIDevice:: openDevice failed to open device. 

Would be nice to know what the workflow is to contribute, or what steps to take to in building Ctrlr. Do we need Introjucer/ Projucer? What settings in 2019? And so on. Because when I try to compile it, VS throws about 180 errors my way. So there goes adding any code or contribution.

**Comment by Possemo** [comment]
I am a C++ noob so my advices may be taken with caution. I could build Ctrlr sucessfully with vs2019 but I don't know if I did it correctly. I have to say that if you are not really a C++ wizard you are not likely to contribute anything useful. This looks damn complex to me. Think about that there are used several 3rd party libraries that have to work together. But if you want to build it yourself (on Windows) here is what I did in short:
- download zipfile from github and unpack somewhere
- decompress the folder ..\ctrlr-master\Source\Misc\boost - This is the boost library which is needed to build Ctrlr
- copy ..\ctrlr-master\Packaging\Windows\icon.ico to ..\ctrlr-master\Builds\VisualStudio2019 - don't ask me why this has to be done
- If you have installed "vs2019 community" you can doubleclick ..\ctrlr-master\Builds\VisualStudio2019\Ctrlr.sln - this will open the vs2019 solution
- If you don't want to change paths copy the newest JUCE library from here [https://juce.com/get-juce/download](https://juce.com/get-juce/download) and unpack it to C:\devel\ctrlr\JUCE\

I installed vs2019 with the following options. Maybe there are some useless options, but at least it seems like there is nothing missing in this config:
Web and Cloud: ASP.NET and Webdev / Python-Dev / Node.js Dev
Desktop & Mobile: .NET Desktopdev / Desktopdev with C++ / Dev for universal Windows platform
Gaming: Gamedev with C++

**Reply by Ted-juh** [reply]
Thank you for the explanation, Possemo. Appreciated. A first test resulted in dramatically fewer errors. Still some "no such file or directory" errors for juce_audio basics.h. Strangely enough, it is there in the modules folder. Going to give it another go at it tonight/ tomorrow.

**Reply by shooking** [reply]
I think the problem _could_ be fixed if @RomanKubiak mandated we build in say c:\Ctrlr or some hard wired directory.
The Jucer has(d) ..\ paths and new Juce insists on a fixed path.

c:\Ctrlr seems like a good anchor  point to git clone to ie
cd c:\
git clone ...ctrlr

then we ALL have a consistent place to build
On the Linux we can ln -s to anywhere.

My 2 cents.

**Comment by spiffo-smith** [comment]
OK, I downloaded the pre-built version of 5.6 and hit a fairly fundamental problem straightaway. It won't connect to my Roland Quad Capture MIDI Input Port. Tried rebooting but no joy, I can't get a MIDI Input setup!

**Reply by damiensellier** [reply]
5.6.0 is a pre relese and midi i/o do not work properly.
Please dowload 5.5.9 instead https://github.com/RomanKubiak/ctrlr/releases/tag/5.5.9

**Reply by Godlike-Productions** [reply]
Or 5.6.23 which does support MIDI, though is still a bit buggy. It can crash without showing error messages. If you want VST3's, 5.6.23 is really the only option.
